webrender/src/frame_builder.rs

```diff
@@ -793,6 +792,59 @@ impl FrameBuilder {
         }
     }
 
+    pub fn add_box_shadow_no_blur(&mut self,
+                                  scroll_layer_id: ScrollLayerId,
+                                  box_bounds: &LayerRect,
+                                  clip_region: &ClipRegion,
+                                  box_offset: &LayerPoint,
+                                  color: &ColorF,
+                                  spread_radius: f32,
+                                  border_radius: f32,
+                                  clip_mode: BoxShadowClipMode) {
+        assert!(spread_radius == 0.0); // TODO: fix cases where this isn't true.
+        let bs_rect = box_bounds.translate(box_offset);
+
+        // We can draw a rectangle instead with the proper border radius clipping.
+        let mut extra_clips = Vec::new();
+
+        let bs_clip_mode = match clip_mode {
+            BoxShadowClipMode::Outset | BoxShadowClipMode::None => ClipMode::Clip,
+            BoxShadowClipMode::Inset => ClipMode::ClipOut,
+        };
+
+        let box_clip_mode = match clip_mode {
+            BoxShadowClipMode::Outset | BoxShadowClipMode::None => ClipMode::ClipOut,
+            BoxShadowClipMode::Inset => ClipMode::Clip,
+        };
+
+        // Clip the inside
+        extra_clips.push(ClipSource::Complex(bs_rect,
```

Comment:
if `add_primitive` accepts `&[ClipSource]`, you could just have a fixed-size array here:
`let extra_clips = [ClipSource::Complex(...), ClipSource::Complex(...)];`